driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -328,4 +332,16 @@ protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out)
             }
         }
     }
+
+    @ChannelHandler.Sharable
+    static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+            ProtocolVersion protocolVersion = frame.header.version;
+            // Check for the protocol version, if the downgrade has occurred during renegotiation
+            if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
+                frame.header.flags.add(Header.Flag.USE_BETA);
```

Comment:
> Possibly related to this: how does auto-negotiation behave if

This is what we currently have. The initial patch was adding `beta` to the protocol version, although we discussed and decided to go the protocol-negotiation path. 

> outgoing frames did not contain the beta flag and we add it on the fly

That was a bit easier than having ProtocolOptions (or that flag) passed through to `ProtocolEncoder` (since it's currently static). 

What we can do is adding a "soft" coding to fix both issues. If we know that driver supports v5 only in beta, we'd throw a configuration exception in cluster builder. Same simplification can be done in `ProtocolEncoder`, where we can just hardwire beta flag for all v5 messages on creation instead of having it added dynamically.

cc @adutra